### PR TITLE
esp32: drivers: flash: add main flash support

### PR DESCRIPTION
**Pull request by glaubermaroto**
Add support for the ESP32 main SPI flash chip (the same SPI flash chip from which program runs).

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

### CODE REVIEW

File: drivers/flash/flash_esp32.c [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+
+			prog_len += page_size;
+		}
+
+		/* remain parts to program */
+		rc = program_page(dev,
+				  address + prog_len,
+				  (uint32_t *)buffer + (prog_len >> 2),
+				  length - prog_len);
+		if (rc) {
```
Comment (sylvioalves) [inline drivers/flash/flash_esp32.c:480]: Is this if necessary?

File: drivers/flash/flash_esp32.c [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+		spi_flash_ll_set_address(hw, start_addr & ADDRESS_MASK_24BIT);
+		spi_flash_ll_erase_sector(hw);
+
+		rc = flash_esp32_wait_cmd_done(hw);
+		if (rc) {
+			return rc;
+		}
+
+		rc = wait_idle(dev);
+		if (rc) {
```
Comment (sylvioalves) [inline drivers/flash/flash_esp32.c:511]: Is this if necessary?

File: drivers/flash/flash_esp32.c (hunk 1 above)
Comment (glaubermaroto) [reply drivers/flash/flash_esp32.c:480]: I've added a `LOG_ERR` message to it. I guess it is more meaningful now.

File: drivers/flash/flash_esp32.c (hunk 2 above)
Comment (glaubermaroto) [reply drivers/flash/flash_esp32.c:511]: I've added a `LOG_ERR` message to it. I guess it is more meaningful now.

**Review by andyross — APPROVED** [review]
Looks clean to me, though I'm not an expert on the SPI layer.

**Review by sylvioalves — APPROVED** [review]

**Pull request merged**